# Práctica Storytelling

* Aplicar los principios y recomendaciones de visualización discutidas en este módulo con la finalidad de hacer una presentación efectiva de resultados.

**Paso a paso:

* Considera la información contenida en la siguiente página Web de Naciones Unidas (http://data.un.org):
* Desarrolla una historia sobre alguna de las áreas que ahí vienen enlistadas (Population, National Acccounts, etc.).
* Usa las recomendaciones de visualización explicadas en este módulo para gráficos (como Gestalt, Tufte y Holmes). 
* Sé claro en sus explicaciones, haciendo anotaciones finales en su notebook de trabajo que indiquen los principios utilizados.

In [121]:
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
warnings.filterwarnings('ignore')

TIP 1. Entiende a la Audicencia

In [141]:
# carga de datos
df = pd.read_csv('C:/Users/Isaac/Desktop/IHD/EBAC DT/CIENCIA DE DATOS/M57 DS/Indicators.csv', encoding = 'Latin1', skiprows = 1)
df

Region/Country/Area                     Unnamed: 1  Year  \
0                       1  Total, all countries or areas  2010   
1                       1  Total, all countries or areas  2010   
2                       1  Total, all countries or areas  2010   
3                       1  Total, all countries or areas  2010   
4                       1  Total, all countries or areas  2010   
...                   ...                            ...   ...   
6751                  722                           SIDS  2022   
6752                  722                           SIDS  2022   
6753                  722                           SIDS  2022   
6754                  722                           SIDS  2022   
6755                  722                           SIDS  2022   

                                                 Series Value  \
0          Population annual rate of increase (percent)   1.3   
1             Total fertility rate (children per women)   2.6   
2     Infant mortality for both sexes (per 1,000 liv...  37.1   
3     Maternal mortality ratio (deaths per 100,000 p...   254   
4       Life expectancy at birth for both sexes (years)  70.1   
...                                                 ...   ...   
6751          Total fertility rate (children per women)   2.3   
6752  Infant mortality for both sexes (per 1,000 liv...  27.3   
6753    Life expectancy at birth for both sexes (years)  72.5   
6754         Life expectancy at birth for males (years)  69.6   
6755       Life expectancy at birth for females (years)  75.5   

                                           Footnotes  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
...                                              ...   
6751  Projected estimate (medium fertility variant).   
6752  Projected estimate (medium fertility variant).   
6753  Projected estimate (medium fertility variant).   
6754  Projected estimate (medium fertility variant).   
6755  Projected estimate (medium fertility variant).   

                                                 Source  
0     United Nations Population Division, New York, ...  
1     United Nations Population Division, New York, ...  
2     United Nations Statistics Division, New York, ...  
3     World Health Organization (WHO), the United Na...  
4     United Nations Population Division, New York, ...  
...                                                 ...  
6751  United Nations Population Division, New York, ...  
6752  United Nations Statistics Division, New York, ...  
6753  United Nations Population Division, New York, ...  
6754  United Nations Population Division, New York, ...  
6755  United Nations Population Division, New York, ...  

[6756 rows x 7 columns]

In [142]:
valores = df['Series'].unique()
valores

array(['Population annual rate of increase (percent)',
       'Total fertility rate (children per women)',
       'Infant mortality for both sexes (per 1,000 live births)',
       'Maternal mortality ratio (deaths per 100,000 population)',
       'Life expectancy at birth for both sexes (years)',
       'Life expectancy at birth for males (years)',
       'Life expectancy at birth for females (years)'], dtype=object)

In [143]:
# trabajaremos con 'Life expectancy at birth for males (years)' del año 2022
# creamos un nuevo df con la columna 'Series' filtrando por el campo 'Life expectancy at birth for males (years)' 

df_exmales = df[df['Series'] == 'Life expectancy at birth for males (years)']
df_exmales.head()

Region/Country/Area                     Unnamed: 1  Year  \
5                     1  Total, all countries or areas  2010   
12                    1  Total, all countries or areas  2015   
19                    1  Total, all countries or areas  2020   
25                    1  Total, all countries or areas  2022   
32                    2                         Africa  2010   

                                        Series Value  \
5   Life expectancy at birth for males (years)  67.6   
12  Life expectancy at birth for males (years)  69.3   
19  Life expectancy at birth for males (years)  69.4   
25  Life expectancy at birth for males (years)  69.1   
32  Life expectancy at birth for males (years)  57.0   

                                         Footnotes  \
5                                              NaN   
12                                             NaN   
19                                             NaN   
25  Projected estimate (medium fertility variant).   
32                                             NaN   

                                               Source  
5   United Nations Population Division, New York, ...  
12  United Nations Population Division, New York, ...  
19  United Nations Population Division, New York, ...  
25  United Nations Population Division, New York, ...  
32  United Nations Population Division, New York, ...

In [144]:
# filtramos los valores que sean igual a 'Africa'
df_exmales = df[df['Series'] == 'Life expectancy at birth for males (years)']
df_exmales = df_exmales[df_exmales['Unnamed: 1'] == 'Africa']
df_exmales

Region/Country/Area Unnamed: 1  Year  \
32                    2     Africa  2010   
39                    2     Africa  2015   
46                    2     Africa  2020   
52                    2     Africa  2022   

                                        Series Value  \
32  Life expectancy at birth for males (years)  57.0   
39  Life expectancy at birth for males (years)  59.3   
46  Life expectancy at birth for males (years)  60.3   
52  Life expectancy at birth for males (years)  60.2   

                                         Footnotes  \
32                                             NaN   
39                                             NaN   
46                                             NaN   
52  Projected estimate (medium fertility variant).   

                                               Source  
32  United Nations Population Division, New York, ...  
39  United Nations Population Division, New York, ...  
46  United Nations Population Division, New York, ...  
52  United Nations Population Division, New York, ...

In [145]:
df_exmales = df_exmales[['Unnamed: 1','Value','Year']]
df_exmales.rename({'Unnamed: 1':'Country', 'Value':'Life expectancy at birth for males'}, axis = 1, inplace = True)
df_exmales = df_exmales.set_index('Country')
df_exmales

Life expectancy at birth for males  Year
Country                                         
Africa                                57.0  2010
Africa                                59.3  2015
Africa                                60.3  2020
Africa                                60.2  2022

In [146]:
# grafico de la espernza de bida la nacer de hombres
gf1 = go.Figure()
gf1.add_trace(go.Scatter(x = df_exmales['Life expectancy at birth for males'] , y = df_exmales['Year'],
                        name = 'Life expectancy', line = dict(color = 'green')))
gf1.update_layout(title = 'Esperanza de vida de los Hombres al Nacer en Africa',
                 xaxis_title = 'Life expectancy Males',
                 yaxis_title = 'Year')
gf1.show()

**La grafica nos indica que la ezperanza de vida en los Hombres para Africa a aumentado 3.2 años del 2010 al 2020**

Tip 2 Usar la Visualización correcta.

In [166]:
# Grafico de dispersión para Population annual rate of increase (percent)
df_rateinc = df[df['Series'] == 'Population annual rate of increase (percent)']
df_rateinc = df_rateinc[df_rateinc['Unnamed: 1'] == 'Africa']
df_rateinc

Region/Country/Area Unnamed: 1  Year  \
27                    2     Africa  2010   
34                    2     Africa  2015   
41                    2     Africa  2020   
48                    2     Africa  2022   

                                          Series Value  \
27  Population annual rate of increase (percent)   2.6   
34  Population annual rate of increase (percent)   2.6   
41  Population annual rate of increase (percent)   2.4   
48  Population annual rate of increase (percent)   2.3   

                                         Footnotes  \
27                                             NaN   
34                                             NaN   
41                                             NaN   
48  Projected estimate (medium fertility variant).   

                                               Source  
27  United Nations Population Division, New York, ...  
34  United Nations Population Division, New York, ...  
41  United Nations Population Division, New York, ...  
48  United Nations Population Division, New York, ...

In [167]:
df_rateinc = df_rateinc[['Unnamed: 1','Value','Year']]
df_rateinc.rename({'Unnamed: 1':'Country', 'Value':'Population annual rate of increase (percent)'}, axis = 1, inplace = True)
df_rateinc = df_rateinc.set_index('Country')
df_rateinc

Population annual rate of increase (percent)  Year
Country                                                   
Africa                                           2.6  2010
Africa                                           2.6  2015
Africa                                           2.4  2020
Africa                                           2.3  2022

In [177]:
gf2 = go.Figure()
gf2.add_trace(go.Scatter(x = df_rateinc['Population annual rate of increase (percent)'], y = df_rateinc['Year'], 
                          mode = 'markers', name = 'rate annual',
                          line = dict(color = 'red')))
gf2.update_layout(title = 'Incremento de la Población anual en Africa',
xaxis_title = 'Incremento de la Poblacional en Africa %',
yaxis_title = 'Año')
gf2.show()

**La gráfica nos muestra un incremento de la población en Africa del 2010 al 2022**

Tip 3. Enfatiza puntos clave

In [183]:
# Gráfico de barras para Maternal mortality ratio (deaths per 100,000 population)
df_ratiomor = df[df['Series'] == 'Maternal mortality ratio (deaths per 100,000 population)']
df_ratiomor = df_ratiomor[df_ratiomor['Unnamed: 1'] == 'Africa']
df_ratiomor

Region/Country/Area Unnamed: 1  Year  \
30                    2     Africa  2010   
37                    2     Africa  2015   
44                    2     Africa  2020   

                                               Series Value Footnotes  \
30  Maternal mortality ratio (deaths per 100,000 p...   593       NaN   
37  Maternal mortality ratio (deaths per 100,000 p...   526       NaN   
44  Maternal mortality ratio (deaths per 100,000 p...   487       NaN   

                                               Source  
30  World Health Organization (WHO), the United Na...  
37  World Health Organization (WHO), the United Na...  
44  World Health Organization (WHO), the United Na...

In [184]:
df_ratiomor = df_ratiomor[['Unnamed: 1','Value','Year']]
df_ratiomor.rename({'Unnamed: 1':'Country', 'Value':'Maternal mortality ratio (deaths per 100,000 population)'}, 
                   axis = 1, inplace = True)
df_ratiomor = df_ratiomor.set_index('Country')
df_ratiomor

Maternal mortality ratio (deaths per 100,000 population)  Year
Country                                                               
Africa                                                 593        2010
Africa                                                 526        2015
Africa                                                 487        2020

In [189]:
gf3 = go.Figure()
gf3.add_trace(go.Bar(x = df_ratiomor['Maternal mortality ratio (deaths per 100,000 population)'], y = df_ratiomor['Year'],
                     name = 'Ratio Mortality',
                     marker = dict(color = 'blue')))
gf3.update_layout(title = 'Relación de Mortalidad Maternal en Africa',
xaxis_title = 'Deaths',
yaxis_title = 'Africa deaths per 100,000 population')
gf3.show()

**La gráfica nos indica que la mortalidad maternal ha disminuido en Africa, para el año 2015 se registraron 593 decesos, para el año 2020 se registraron 487 decesos.**

Tip 4 Cuenta una historia con tus datos

In [220]:
# Filtrar los datos para 'Life expectancy at birth for males' y 'females' en África
df_morinf = df[(df['Series'] == 'Life expectancy at birth for males (years)') | 
               (df['Series'] == 'Life expectancy at birth for females (years)')]
df_morinf = df_morinf[df_morinf['Unnamed: 1'] == 'Africa']

# Ajustar las columnas y renombrar
df_morinf = df_morinf[['Series', 'Value', 'Year']]
df_morinf.rename({'Value': 'Life Expectancy'}, axis=1, inplace=True)

# Pivotar la tabla para obtener columnas separadas para hombres y mujeres
df_pivot = df_morinf.pivot_table(values='Life Expectancy', index='Year', columns='Series')

# Crear la gráfica apilada
gf4 = go.Figure()

# Agregar traza para hombres
gf4.add_trace(go.Scatter(
    x=df_pivot.index,
    y=df_pivot['Life expectancy at birth for males (years)'],
    mode='lines',
    name='Males',
    stackgroup='one'
))

# Agregar traza para mujeres
gf4.add_trace(go.Scatter(
    x=df_pivot.index,
    y=df_pivot['Life expectancy at birth for females (years)'],
    mode='lines',
    name='Females',
    stackgroup='one'
))

# Configurar el layout del gráfico
gf4.update_layout(
    title='Life Expectancy in Africa (Males vs Females)',
    xaxis_title='Year',
    yaxis_title='Life Expectancy (years)',
    showlegend=True
)

# Mostrar la gráfica
gf4.show()


**Gráfica apilada que muestra la expectativa de vida tanto para hombres como para mujeres a lo largo de los años en África.**

Tip 5 Creación de Cuadro de Control (Dashboard) para los gráficos previos.

In [222]:
import dash
from dash import dcc
from dash import html

# consolidación de graficos
app = dash.Dash(__name__)
app.layout = html.Div([
    html.H1('Cuadro de Control de Mando(Dashboard)'),
    
    html.Div([
        dcc.Graph(figure = gf1),
        dcc.Graph(figure = gf2),
    ], style = {'display':'flex'}),
    
    html.Div([
        dcc.Graph(figure = gf3),
        dcc.Graph(figure = gf4),
    ], style = {'display':'flex'}),
])
if __name__== '__main__':
    app.run_server(debug = True, port = 8051)